### PR TITLE
fix(externals): mismatched externals caused one-app to fail to start

### PR DESCRIPTION
**Pull request by Matthew-Mallimo**
A child module that requires an external that is NOT provided by its root module will cause one-app to fail to start up, or fail polling if already running.

## Description

This fix works by catching an error that is thrown by `onModuleLoad`, a function passed to holocron from the one-app server. Typically this error would bubble up to one-app, where the module map polling would fail. 

By catching the error, we are able to ignore the "bad" module by preventing it from being updated in the in-memory module map.

This means that if an existing module is updated to require an external that is NOT provided by the parent, we will use its last stable version until a working version of that module is deployed.

This does not affect the module map polling. It will continue to climb as normal up to the ONE_MAP_POLLING_MAX value. 

## Motivation and Context

This can prevent developers from starting the one-app server locally if a faulty module is deployed to the environment their one-app-runner config points to. It also prevents other modules from being updated in the module map until the faulty module is corrected.

## How Has This Been Tested?

Pending unit tests

## Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (adding or updating documentation)
- [ ] Dependency update

## Checklist:

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] My changes are in sync with the code style of this project.
- [ ] There aren't any other open Pull Requests for the same issue/update.
- [ ] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] This change impacts caching for client browsers.
- [ ] This change adds additional environment variable requirements for Holocron users.
- [ ] I have added the Apache 2.0 license header to any new files created.

### CODE REVIEW

File: packages/holocron/src/updateModuleRegistry.js [hunk 1]
```diff
@@ -35,32 +35,62 @@ function defaultGetModulesToUpdate(curr, next) {
...
+        try {
+          const loadedModule = await loadModule(
+            moduleName,
+            unsanitizedModuleMap.modules[moduleName],
+            onModuleLoad
+          );
+          return addHigherOrderComponent(loadedModule);
+        } catch (e) {
+          // eslint-disable-next-line no-console
+          console.error(e);
```
Comment (10xLaCroixDrinker) [inline packages/holocron/src/updateModuleRegistry.js:62]: Add some details/context to this error for the users

**Review by 10xLaCroixDrinker — CHANGES_REQUESTED** [review]

File: packages/holocron/src/updateModuleRegistry.js (hunk 1 above)
Comment (JAdshead) [reply packages/holocron/src/updateModuleRegistry.js:62]: I don't want to bloat this PR but we should consider adding URLs to our error messages

File: packages/holocron/src/updateModuleRegistry.js [hunk 2]
```diff
@@ -35,32 +35,62 @@ function defaultGetModulesToUpdate(curr, next) {
...
+  const sanitizedModuleMap = { ...unsanitizedModuleMap };
+  // Keep working version of modules if they are in the list of problem modules
+  problemModules.forEach((module) => {
+    if (currentModuleMap.modules) {
+      sanitizedModuleMap.modules[module] = currentModuleMap.modules[module];
+    } else {
+      // If it doesn't exist in the old module map, that means its being deployed for the first time
+      // or holocron is initializing modules for the first time. If there is an issue with the
+      // module we should exclude it from the module map entirely.
+      delete sanitizedModuleMap.modules[module];
```
Comment (JAdshead) [inline packages/holocron/src/updateModuleRegistry.js:89]: can we add a test for this scenario and the inverse

File: packages/holocron/__tests__/updateModuleRegistry.spec.js [hunk 3]
```diff
@@ -525,11 +524,63 @@ describe('updateModuleRegistry', () => {
...
     };
     const batchModulesToUpdate = (x) => x.map((i) => [i]);
-    await expect(updateModuleRegistry({
+    const updatedRegistry = await updateModuleRegistry({
+      moduleMap: newModuleMap,
+      onModuleLoad,
+      batchModulesToUpdate,
+    });
+    expect(updatedRegistry['another-module']).toBeTruthy();
+    expect(updatedRegistry['best-module']).toBeFalsy();
```
Comment (JAdshead) [inline packages/holocron/__tests__/updateModuleRegistry.spec.js:533]: we should also test to make sure that the error is still being logged

File: packages/holocron/src/updateModuleRegistry.js (hunk 2 above)
Comment (Matthew-Mallimo) [reply packages/holocron/src/updateModuleRegistry.js:89]: The "should not throw if any of the modules fail to load - empty module map" test case covers this scenario

**Review by infoxicator — COMMENTED** [review]
LGTM, we still need to add integration tests for the new scenarios

File: packages/holocron/src/updateModuleRegistry.js [hunk 4]
```diff
@@ -58,8 +58,15 @@ export default async function updateModuleRegistry({
...
-          // eslint-disable-next-line no-console
-          console.error(e);
+          const brokenUrl = unsanitizedModuleMap.modules[moduleName].node.url;
+          if (currentModuleMap.modules && currentModuleMap.modules[moduleName]) {
+            const previousUrl = currentModuleMap.modules[moduleName].node.url;
+            // eslint-disable-next-line no-console
+            console.error(`There was an error loading module ${moduleName} at ${brokenUrl}. Reverting back to ${previousUrl}`, e);
+          } else {
+            // eslint-disable-next-line no-console
+            console.error(`There was an error loading module ${moduleName} at ${brokenUrl}. Removing ${moduleName} from in-memory module map until next module map poll.`, e);
```
Comment (JAdshead) [inline packages/holocron/src/updateModuleRegistry.js:68]: Could the implementation detail add confusion 

```suggestion
            console.error(`There was an error loading module ${moduleName} at ${brokenUrl}. Ignoring ${moduleName} until next module map poll.`, e);
```

File: packages/holocron/__tests__/updateModuleRegistry.spec.js [hunk 5]
```diff
@@ -482,9 +483,9 @@ describe('updateModuleRegistry', () => {
     expect(updatedModules).toMatchSnapshot();
   });
 
-  it('should throw if any of the modules fail to load', async () => {
-    expect.assertions(1);
+  it('should not throw if any of the modules fail to load', async () => {
     const mockLoadModule = async (moduleName, moduleVersion) => `new ${moduleName}@${moduleVersion}`;
+    console.error = jest.fn();
```
Comment (10xLaCroixDrinker) [inline packages/holocron/__tests__/updateModuleRegistry.spec.js:488]: Consider:

```suggestion
    jest.spyOn(console, 'error').mockImplementation(x => x);
```

**Review by 10xLaCroixDrinker — APPROVED** [review]

**Review by 10xLaCroixDrinker — APPROVED** [review]

**Review by JAdshead — APPROVED** [review]

**Review by code-forger — APPROVED** [review]

**Pull request merged**